### PR TITLE
Refactor save/restore concept: Worker dimension as outer dimension

### PR DESCRIPTION
**Pull request by philipp-fischer**
This PR changes the way how the internal dataset wrappers save and restore their states.
Instead of every dataset keeping the states for all workers, they now only keep the state of the current worker.
Also, the checkpoint format changes. The saved state will contain a list of dataset states, one for each worker.
Before, the worker dimension was an inner dimension of the state of each dataset.

This will allow us to re-distribute the states easily, when resuming with a different node count but same number of global workers.

### CODE REVIEW

File: src/megatron/energon/flavors/webdataset/joined_webdataset.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ def build(self, worker_rotation_offset: int = 0) -> SavableDataset[T_sample]:
 
         dataset = WebdatasetSampleLoaderDataset(
             join_readers=itar_readers,
-            workers_sample_slice_offsets=workers_sample_slice_offsets,
+            sample_slice_offsets=workers_sample_slice_offsets,
```
Comment (voegtlel) [inline src/megatron/energon/flavors/webdataset/joined_webdataset.py:189]: Why rename without `workers`?

File: src/megatron/energon/flavors/webdataset/sample_loader.py [hunk 2]
```diff
@@ -35,60 +24,23 @@ class RawSampleData:
...
 
 class WebdatasetSampleLoaderDataset(SavableDataset[RawSampleData]):
     """Internal class for loading samples from webdataset slices"""
 
     #: The readers for each joined dataset
     join_readers: Sequence[ITarReader]
+
     #: The offsets of the slice slices to iterate over
-    worker_slice_offsets: Sequence[Sequence[int]]
+    slice_offsets: Sequence[int]
```
Comment (voegtlel) [inline src/megatron/energon/flavors/webdataset/sample_loader.py:43]: Naming is now inconsistent? I see usage of both, `workers_slice_offsets` and `slice_offsets` now. But I guess `slice_offsets` should actually be local to iter?

File: src/megatron/energon/savable_loader.py [hunk 3]
```diff
@@ -485,37 +442,25 @@ def _save_state(self) -> SavableDatasetState:
...
-            sample_index=[0 if s is None else s.sample_index for s in states],
-        )
-
-    def _restore_state(self, state: SavableDatasetMergedState) -> None:
+    def _restore_state(self, state: SavableDatasetState) -> None:
         """Restores the internal worker state"""
         assert torch.utils.data.get_worker_info() is not None, "Can only restore in worker process"
-        if state.torch_rng[self._worker_id] is None:
+        if state.torch_rng is None:
             torch.manual_seed(torch.initial_seed())
```
Comment (voegtlel) [inline src/megatron/energon/savable_loader.py:449]: Just seeing this. Is this correct? Shouldn't that rely on the configured seed nowadays? 😅

File: src/megatron/energon/savable_loader.py [hunk 4]
```diff
@@ -546,21 +491,9 @@ def get_checkpoint(self, last_sample_indexes: List[int]) -> SavableDatasetCheckp
...
-        return SavableDatasetMergedCheckpoint(
-            dataset_state=self.dataset.merge_states([c.dataset_state for c in checkpoints]),
-            state=self._merge_states([c.state for c in checkpoints]),
-            offset=[c.offset for c in checkpoints],
-        )
-
     def restore_checkpoint(
         self,
-        state: Optional[SavableDatasetMergedCheckpoint],
+        state_list: Optional[List[SavableDatasetCheckpoint]],
```
Comment (voegtlel) [inline src/megatron/energon/savable_loader.py:496]: Docs are not adapted. Also, why `state` to `state_list`? I think `list` does not help to understand that better (and typing already makes clear, it's a list :D ). Rename e.g. to `workers_state`.

File: src/megatron/energon/wrappers/base.py [hunk 5]
```diff
@@ -25,132 +20,112 @@
...
 
+    def reset_state_deep(self) -> None:
+        """Resets the state of the inner datasets and then the own state."""
 
-class SampleIndex:
+        for ds in self.datasets:
+            if isinstance(ds, BaseWrapperDataset):
+                ds.reset_state_deep()
+            else:
+                ds.reset_state_own()
```
Comment (voegtlel) [inline src/megatron/energon/wrappers/base.py:98]: Why `own` if deep is requested?

File: src/megatron/energon/wrappers/base.py [hunk 6]
```diff
@@ -25,132 +20,112 @@
...
-        res = self._sample_index[self._rank]
-        self._sample_index[self._rank] += 1
+        res = self._sample_index
+        self._sample_index += 1
         return res
 
     @property
     def current_idx(self) -> int:
-        return self._sample_index[self._rank]
+        return self._sample_index
```
Comment (voegtlel) [inline src/megatron/energon/wrappers/base.py:128]: Could directly rename the variable to `current_idx` and expose it, instead of using `@property`

File: src/megatron/energon/wrappers/batch_dataset.py [hunk 7]
```diff
@@ -104,9 +75,13 @@ def __init__(
         self.batcher_config = batcher_config
         self.drop_last = drop_last
         self.error_handler = error_handler
-        self._sample_index = SampleIndex(worker_config, src=self)
-        self._generator_sample_keys = [None] * max(self.worker_config.num_workers, 1)
-        self._generator_offsets = [None] * max(self.worker_config.num_workers, 1)
+
+        self.reset_state_own()
```
Comment (voegtlel) [inline src/megatron/energon/wrappers/batch_dataset.py:79]: In general, I wonder if `reset_state_own` should only be called in the worker? I.e. not in the init at all. Because the state actually only exists in the worker, right?

File: src/megatron/energon/wrappers/group_batch_dataset.py [hunk 8]
```diff
@@ -19,68 +19,51 @@
...
 T_batch = TypeVar("T_batch", covariant=True)
 T_batch_sample = TypeVar("T_batch_sample", covariant=True)
 
 
-@dataclass_slots
-class BucketState(State):
-    key: Hashable
-    batch_size: int
-    samples: SampleBufferState
+# @dataclass_slots
```
Comment (voegtlel) [inline src/megatron/energon/wrappers/group_batch_dataset.py:34]: Remove commented code

File: src/megatron/energon/wrappers/shuffle_buffer_dataset.py [hunk 9]
```diff
@@ -51,8 +31,12 @@ def __init__(
         """Create a shuffle buffer for the dataset."""
         super().__init__(dataset, worker_config=worker_config)
         self.size = size
+        self.reset_state_own()
+
+    def reset_state_own(self) -> None:
         self._worker_rng = WorkerRng(self.worker_config)
-        self._active_buffer = SavableSampleBuffer(dataset, worker_config=worker_config)
+        # TODO: Former reset was via restore_state(None). Is this ok too?
```
Comment (voegtlel) [inline src/megatron/energon/wrappers/shuffle_buffer_dataset.py:38]: I guess it would be nice to have this pulled through every class, that the savables are mostly instantiated in the init? And the respective savable will call `.reset_state_*` on the inner types?

**Review by voegtlel — CHANGES_REQUESTED** [review]

File: src/megatron/energon/flavors/base_dataset.py [hunk 10]
```diff
@@ -281,48 +244,72 @@ class SavableDataset(IterableDataset[T_sample], Generic[T_sample], ABC):
 
     worker_config: WorkerConfig
 
+    #: List of names of the fields that are saved and restored in the state.
+    _savable_fields: ClassVar[List[str]] = []
```
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/flavors/base_dataset.py:248]: I looked up the convention an double underscore is for builtin stuff. This is just "protected" or "private".
I can make it a tuple

File: src/megatron/energon/flavors/base_dataset.py [hunk 11]
```diff
@@ -281,48 +244,72 @@ class SavableDataset(IterableDataset[T_sample], Generic[T_sample], ABC):
...
+                getattr(self, key).restore_state(value)
+            else:
+                setattr(self, key, value)
+
+    @abstractmethod
+    def reset_state_own(self) -> None:
+        """Resets the state of the dataset to the initial state. Can only be called in a worker process."""
         ...
 
+    def reset_state_deep(self) -> None:
```
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/flavors/base_dataset.py:309]: Resetting the own state *must* be overridden by the derived classes, but recursion is handled by the base class

File: src/megatron/energon/savable_loader.py (hunk 3 above)
Comment (philipp-fischer) [reply src/megatron/energon/savable_loader.py:449]: Yeah, but may be dangerous to remove if some legacy user code relies on it.

File: src/megatron/energon/wrappers/base.py (hunk 5 above)
Comment (philipp-fischer) [reply src/megatron/energon/wrappers/base.py:100]: The recursion ends when we don't hit any `BaseWrapperDataset` anymore. A normal savable dataset won't have a method `reset_state_deep`, because it does not contain any inner datasets.

File: src/megatron/energon/wrappers/base.py (hunk 6 above)
Comment (philipp-fischer) [reply src/megatron/energon/wrappers/base.py:130]: Done

File: src/megatron/energon/wrappers/group_batch_dataset.py (hunk 8 above)
Comment (philipp-fischer) [reply src/megatron/energon/wrappers/group_batch_dataset.py:34]: Done

File: src/megatron/energon/flavors/webdataset/joined_webdataset.py (hunk 1 above)
Comment (philipp-fischer) [reply src/megatron/energon/flavors/webdataset/joined_webdataset.py]: You spotted a bug

File: src/megatron/energon/flavors/webdataset/sample_loader.py (hunk 2 above)
Comment (philipp-fischer) [reply src/megatron/energon/flavors/webdataset/sample_loader.py]: I fixed a bug around this code, but the naming is still the same. 
We have `self.workers_slice_offsets` which is set in the main process for all workers, and we have 
```
#: The offsets of the slice slices to iterate over for the current worker
slice_offsets: Optional[Sequence[int]]
```

File: src/megatron/energon/wrappers/base.py [hunk 12]
```diff
@@ -25,133 +20,123 @@
...
 
     def assert_can_restore(self) -> None:
-        self.dataset.assert_can_restore()
+        for ds in self.datasets:
+            ds.assert_can_restore()
+
+    def worker_has_samples(self) -> bool:
+        return any(ds.worker_has_samples() for ds in self.datasets)
+
+    def find_wrapped_dataset(self, cls: Type[SavableDataset]) -> Optional[SavableDataset]:
```
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/wrappers/base.py:67]: Should not break at blend actually. It should go into the inner datasets

File: src/megatron/energon/savable_loader.py [hunk 13]
```diff
@@ -925,6 +851,22 @@ def _worker_command(self, *cmd_args) -> List[Any]:
...
+    def get_batch_size(self) -> Optional[int]:
+        """Try to infer micro batch size from the dataset"""
+        if isinstance(self.dataset, SavableDatasetWrapper):
+            dataset = self.dataset.dataset
+        else:
+            dataset = self.dataset
+
+        if (
+            isinstance(dataset, BaseWrapperDataset)
+            and (bds := dataset.find_wrapped_dataset(BatchDataset)) is not None
```
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/savable_loader.py:863]: Yes, I checked it. Starting from Python 3.8

File: src/megatron/energon/flavors/base_dataset.py [hunk 14]
```diff
@@ -281,48 +244,76 @@ class SavableDataset(IterableDataset[T_sample], Generic[T_sample], ABC):
...
+        for key in self._savable_fields:
+            attr = getattr(self, key)
+            if isinstance(attr, Savable):
+                state[key] = attr.save_state()
+            else:
+                # Check if this field is a simple python type or a user class
 
-        Args:
-            states: The states of the workers to merge.
+                if attr is not None and getattr(attr, "__module__", "builtins") != "builtins":
```
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/flavors/base_dataset.py:272]: I didn't see a need. We don't seem to have a use case for that right now.
Do you think we need to check for `Savable` recursively?

File: src/megatron/energon/savable_loader.py [hunk 15]
```diff
@@ -288,8 +210,10 @@ class SavableDatasetWrapper(IterableDataset[Tuple[int, int, T]], Generic[T]):
     _sample_index: int = 0
     _worker_offset: int = 0
     _last_checkpoints: List[SavableCheckpoint]
-    _restore_from: Optional[Any] = None
-    _skip_samples: List[int]
+
+    _workers_restore_from: List[Optional[SavableDatasetState]] = []
```
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/savable_loader.py:214]: Good catch.

File: src/megatron/energon/flavors/base_dataset.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (voegtlel) [reply src/megatron/energon/flavors/base_dataset.py:272]: But this is not for the `Savable`, but for builtins? Savable will check recursively itself?

File: src/megatron/energon/wrappers/base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (voegtlel) [reply src/megatron/energon/wrappers/base.py:67]: And then it would probably return the first found. That would be wrong, because all inner datasets of blend are equal, thus it should return all in parallel, which is not at all needed here. So I still suggest to move this to where it's needed to avoid errors. Also, just using the first doesn't necessarily work.

File: src/megatron/energon/savable_loader.py [hunk 16]
```diff
@@ -288,8 +210,10 @@ class SavableDatasetWrapper(IterableDataset[Tuple[int, int, T]], Generic[T]):
...
+    _workers_dataset_state: List[Optional[FlexState]]
```
[the remark this answers is not in the document]
Comment (voegtlel) [reply src/megatron/energon/savable_loader.py:215]: Not doing this now.

File: src/megatron/energon/wrappers/base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (voegtlel) [reply src/megatron/energon/wrappers/base.py]: Making the method private now. It has it's usefulness for debugging, but shouldn't be exposed.

**Review by voegtlel — APPROVED** [review]

File: src/megatron/energon/savable_loader.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (philipp-fischer) [reply src/megatron/energon/savable_loader.py]: Made private
